tutorials/eth-api/complete-dapp-architecture.md

```diff
@@ -0,0 +1,735 @@
+
+```javascript
+export default function PurchaseOccurredEvents({ contract }) {
+  // Get block number to ensure that the useLogs doesn't search from 0, otherwise it will time out
+  const blockNumber = useBlockNumber();
+
+  // Create a filter & get the logs
+  const filter = { args: [null, null], contract, event: "PurchaseOccurred" };
+  const logs = useLogs(filter, { fromBlock: blockNumber - 10000 });
+  const parsedLogs = logs?.value.slice(-5).map(log => log.data);
+
+  // ... 
+}
+```
+
+Here's what happens in this code:  
+
+1. The block number is received from the `useBlockNumber` hook, similar to using `eth_blockNumber` in the JSON-RPC
+2. A filter is created to filter for all events with any arguments on the contract injected into the component with the event name `PurchaseOccurred`
+3. Logs are queried for via the `useLogs` hook, similar to using `eth_getLogs` in the JSON-RPC. Note that we're only querying the last 10000 blocks, because otherwise the entire history of the blockchain would be queried and the RPC would time-out
+4. The resultant logs are parsed and the most recent 5 are selected
+
+If we want to display them, we can do it like so:  
+
+```javascript
+export default function PurchaseOccurredEvents({ contract }) {
+  // ...
+  return (
+    <Grid item xs={12} marginTop={5}>
+      <TableContainer >
+        <Table>
+          {/* TableHead Component */}
+          <TableBody>
+            {parsedLogs?.reverse().map((log, index) => (
+              <TableRow key={index}>
+                <TableCell>{log.minter}</TableCell>
+                <TableCell align="right">{utils.formatEther(log.amount)} tokens</TableCell>
+              </TableRow>
+            ))}
+          </TableBody>
+        </Table>
+      </TableContainer>
+    </Grid>
+  );
+}
+```
+
+Which can be added to `App.js`:  
+
+```javascript
+import { MintingComponent } from './MintingComponent.js';
+// ... other imports
+
+function App() {
+  // ...
+
+  return (
+    {/* Wrapper Components */}
+      {/* Button Component */}
+      <Card sx={styles.card}>
+        <h1 style={styles.alignCenter}>Mint Your Token!</h1>
+        <SupplyComponent contract={contract} />
+        <MintingComponent contract={contract} />
+        <PurchaseOccurredEvents contract={contract} />
+      </Card>
+    {/* Wrapper Components */}
+  )
+}
+```
+
+And, if you've done any transactions, you'll see that they'll pop up!  
+
+![Finished DApp](/images/tutorials/eth-api/complete-dapp/complete-dapp-5.png)
+
+Now you've implemented 3 main components of DApp frontends: reading from storage, sending transactions, and reading logs. With these building blocks as well as the knowledge you learned with smart contracts and nodes you should be able to cover 80% of DApps. Of course, there are more advanced (but optional) components of DApps that have popped up over time, some of which are listed in the next section.  
+
+## Additional Options {: #additional-options }
```

Comment:
This section (and subsections) seems out of place to me. Also some of this is covered in the Web2 vs Web3 blogpost that should have been published months ago haha.. I'm thinking maybe we can remove it from here and just link to the blogpost so we keep this as more of a code-oriented guide 🤷‍♀️ 

I'll try to figure out when the blogpost will actually be published

Reply:
@eshaben are there parts of here we can include in the Web2 Web3 blogpost?

Reply:
yeah we can always edit the web2 vs web3 blogpost

Reply:
Ok so @eshaben if you feel there is anything here that could add value to the blogpost I would go ahead and edit it - That way, Jeremy's effort were not in vain